### PR TITLE
SONARJAVA-2477 Fix possible infinite recursion when setting constraints on relations

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java [hunk 1]
```diff
@@ -82,6 +82,28 @@ public String toString() {
...
+  public int hashCode() {
+    return 31 * id;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass())
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java]: MINOR Missing curly brace. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00121)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java [hunk 2]
```diff
@@ -82,6 +82,28 @@ public String toString() {
     FALSE_LITERAL
   );
 
+  private static int idGenerator;
+  private int id;
+
+  public SymbolicValue() {
+    id = idGenerator++;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java]: MAJOR Extract this increment or decrement operator into a dedicated statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS881)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 3]
```diff
@@ -170,36 +179,36 @@ RelationalSymbolicValue inverse() {
...
         states.clear();
         states.addAll(newStates);
-      });
-    }
+      }
+    });
     return states;
   }
 
+  private List<ProgramState> applyConstraint(Constraint constraint, SymbolicValue to, List<ProgramState> states, Set<RelationalSymbolicValue> knownRelations) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: MAJOR Make "applyConstraint" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SymbolicValue.java#L186](https://github.com/SonarSource/sonar-java/blob/ce866cf51c916099045ba69cf0ea4e3e4176d5f8/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L186): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L214](https://github.com/SonarSource/sonar-java/blob/ce866cf51c916099045ba69cf0ea4e3e4176d5f8/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L214): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L234](https://github.com/SonarSource/sonar-java/blob/ce866cf51c916099045ba69cf0ea4e3e4176d5f8/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L234): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SymbolicValue.java#L188](https://github.com/SonarSource/sonar-java/blob/e548cd7a3e806ec2c3699554f58d15d3935fb722/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L188): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L216](https://github.com/SonarSource/sonar-java/blob/e548cd7a3e806ec2c3699554f58d15d3935fb722/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L216): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L236](https://github.com/SonarSource/sonar-java/blob/e548cd7a3e806ec2c3699554f58d15d3935fb722/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L236): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java [hunk 4]
```diff
@@ -157,6 +182,10 @@ public ProgramState setSingleConstraint(ProgramState programState, ObjectConstra
     return states.get(0);
   }
 
+  protected List<ProgramState> setConstraint(ProgramState state, Constraint constraint, Set<RelationalSymbolicValue> knownRelations) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java]: MAJOR Remove this unused method parameter "knownRelations". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SymbolicValue.java#L193](https://github.com/SonarSource/sonar-java/blob/06287717bec6d6b4042da11dc472f4b18a33ecbf/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L193): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L221](https://github.com/SonarSource/sonar-java/blob/06287717bec6d6b4042da11dc472f4b18a33ecbf/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L221): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L241](https://github.com/SonarSource/sonar-java/blob/06287717bec6d6b4042da11dc472f4b18a33ecbf/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L241): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 5]
```diff
@@ -102,11 +102,18 @@ public RelationalSymbolicValue(Kind kind) {
 
   @Override
   public List<ProgramState> setConstraint(ProgramState initialProgramState, BooleanConstraint booleanConstraint) {
+    return setConstraint(initialProgramState, booleanConstraint, knownRelations(initialProgramState));
+  }
+
+  @Override
+  protected List<ProgramState> setConstraint(ProgramState initialProgramState, Constraint booleanConstraint, Set<RelationalSymbolicValue> knownRelations) {
+    if (!(booleanConstraint instanceof BooleanConstraint)) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: One option could be to test this against BooleanConstraint.TRUE given that value is tested against FALSE...

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SymbolicValue.java#L193](https://github.com/SonarSource/sonar-java/blob/5655f588098c7aefa5db0f12dca8788acfe71191/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L193): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L221](https://github.com/SonarSource/sonar-java/blob/5655f588098c7aefa5db0f12dca8788acfe71191/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L221): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [SymbolicValue.java#L241](https://github.com/SonarSource/sonar-java/blob/5655f588098c7aefa5db0f12dca8788acfe71191/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java#L241): Override the "equals" method in this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2160)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**